### PR TITLE
fix #5997: refactor function registerRef

### LINKED ISSUE #5997: Refs incorrect when using v-for and combining components and elements
### Version
2.3.4

### Reproduction link
[https://jsfiddle.net/floorish/7td2offb/](https://jsfiddle.net/floorish/7td2offb/)

### Steps to reproduce
- Parent with dynamic components, referenced by `ref`: `<component :is="child.type"  v-for="child in children" :key="child.id" ref="children"></component>`
- Multiple `child.type`s where one of them uses a loading component: `<loading v-if="isLoading"></loading> <div v-else>Child B ready</div>`
- In Child switch `isLoading` from `true` to `false` after a timeout/async call

### What is expected?
`parent.$refs.children` equals the displayed child components

### What is actually happening?
`parent.$refs.children` only includes the child component that has the loading mechanism

---
This is very much an edge case, but it took me a while to figure out any solutions. Most trivial solution is to wrap the Child template in another `<div>`, but AFAIK the `<div v-if="val"></div> <div v-else></div>` construct in the root of a template is supported in Vue (there is always a single root element).

Other solutions are provided in reproduction link.

### PR DESCRIPTION
**Pull request by javoski**
fix #5997 .
When registering `ref`, if `refs[key]` is an array and `ref` is already in `refs[key]`, then do nothing.

**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

### CODE REVIEW

File: src/core/vdom/modules/ref.js [hunk 1]
```diff
@@ -32,10 +32,11 @@ export function registerRef (vnode: VNodeWithData, isRemoval: ?boolean) {
...
   } else {
     if (vnode.data.refInFor) {
-      if (Array.isArray(refs[key]) && refs[key].indexOf(ref) < 0) {
-        refs[key].push(ref)
-      } else {
+      if (!Array.isArray(refs[key])) {
         refs[key] = [ref]
+      } else if (refs[key].indexOf(ref) < 0) {
+        // $flow-disable-line
+        refs[key].push(ref)
```
Comment (javoski) [inline src/core/vdom/modules/ref.js]: It's seems that `flow` is a little weak to check array type, even codes below couldn't pass.
```js
if (Array.isArray(refs[key])) {
  if (refs[key].indexOf(ref)) {
    refs[key].push(ref)
  }
}
```
So I added a suppress_comment option to `.flowconfig`.

**Comment by Kingwl** [comment]
please add the test case😁

**Comment by javoski** [comment]
Will do it soon
wenlu.wang <notifications@github.com>于2017年7月6日 周四19:12写道：

**Comment by wuchuguang** [comment]
upgrade vue2.4.4
parent1
     child ref="child"
         loop child

parent1.$refs.child  is ok
parent2.$refs.child is undefined

**Pull request merged**